parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,843 @@
+    for (int i = 0; i < numBytes; ++i) {
+      bytes[pos + i] = (byte) ((value >>> (8 * i)) & 0xFF);
+    }
+  }
+
+  public static byte primitiveHeader(int type) {
+    return (byte) (type << 2 | PRIMITIVE);
+  }
+
+  public static byte shortStrHeader(int size) {
+    return (byte) (size << 2 | SHORT_STR);
+  }
+
+  public static byte objectHeader(boolean largeSize, int idSize, int offsetSize) {
+    return (byte) (((largeSize ? 1 : 0) << (BASIC_TYPE_BITS + 4))
+        | ((idSize - 1) << (BASIC_TYPE_BITS + 2))
+        | ((offsetSize - 1) << BASIC_TYPE_BITS)
+        | OBJECT);
+  }
+
+  public static byte arrayHeader(boolean largeSize, int offsetSize) {
+    return (byte) (((largeSize ? 1 : 0) << (BASIC_TYPE_BITS + 2)) | ((offsetSize - 1) << BASIC_TYPE_BITS) | ARRAY);
+  }
+
+  /**
+   * Check the validity of an array index `pos`.
+   * @param pos The index to check
+   * @param length The length of the array
+   * @throws MalformedVariantException if the index is out of bound
+   */
+  public static void checkIndex(int pos, int length) {
+    if (pos < 0 || pos >= length) {
+      throw new IllegalArgumentException(
+          String.format("Invalid byte-array offset (%d). length: %d", pos, length));
+    }
+  }
+
+  /**
+   * Reads a little-endian signed long value from `bytes[pos, pos + numBytes)`.
+   * @param bytes The byte array to read from
+   * @param pos The starting index of the byte array to read from
+   * @param numBytes The number of bytes to read
+   * @return The long value
+   */
+  static long readLong(byte[] bytes, int pos, int numBytes) {
+    checkIndex(pos, bytes.length);
+    checkIndex(pos + numBytes - 1, bytes.length);
+    long result = 0;
+    // All bytes except the most significant byte should be unsigned-extended and shifted
+    // (so we need & 0xFF`). The most significant byte should be sign-extended and is handled
+    // after the loop.
+    for (int i = 0; i < numBytes - 1; ++i) {
+      long unsignedByteValue = bytes[pos + i] & 0xFF;
+      result |= unsignedByteValue << (8 * i);
+    }
+    long signedByteValue = bytes[pos + numBytes - 1];
+    result |= signedByteValue << (8 * (numBytes - 1));
+    return result;
+  }
+
+  /**
+   * Read a little-endian unsigned int value from `bytes[pos, pos + numBytes)`. The value must fit
+   * into a non-negative int (`[0, Integer.MAX_VALUE]`).
+   */
+  static int readUnsigned(byte[] bytes, int pos, int numBytes) {
+    checkIndex(pos, bytes.length);
+    checkIndex(pos + numBytes - 1, bytes.length);
+    int result = 0;
+    // Similar to the `readLong` loop, but all bytes should be unsigned-extended.
+    for (int i = 0; i < numBytes; ++i) {
+      int unsignedByteValue = bytes[pos + i] & 0xFF;
+      result |= unsignedByteValue << (8 * i);
+    }
+    if (result < 0) {
+      throw new MalformedVariantException(String.format("Failed to read unsigned int. numBytes: %d", numBytes));
```

Comment:
As with other places, it doesn't always make sense for this to throw `MalformedVariantException` because that assumes how this is called. In order to throw `MalformedVariantException`, this check should be in the calling code that is decoding an offset, rather than here. With the call here, this is violating some other expectation of the method -- that the value will fit in an unsigned int -- even though there is no restriction on `numBytes`.